src/verifiable/credential.ts

```diff
@@ -134,6 +137,128 @@ export class W3CCredential {
+          resolverURL,
+          userDID,
+          credStatusResolverRegistry
+        );
+      }
+      case ProofType.Iden3SparseMerkleTreeProof: {
+        return this.verifyIden3SparseMerkleTreeProof(
+          proof as Iden3SparseMerkleTreeProof,
+          coreClaim,
+          resolverURL
+        );
+      }
+      default: {
+        throw new Error('invalid proof type');
+      }
+    }
+  }
+
+  private async verifyBJJSignatureProof(
+    proof: BJJSignatureProof2021,
+    coreClaim: Claim,
+    resolverURL: string,
+    userDID: DID,
+    credStatusResolverRegistry: CredentialStatusResolverRegistry
+  ): Promise<boolean> {
+    // issuer auth claim
+    const authClaim = proof.issuerData.authCoreClaim;
+    const rawSlotsInt = authClaim.rawSlotsAsInts();
+    const pubKey = new PublicKey([rawSlotsInt[2], rawSlotsInt[3]]);
+
+    // core claim hash
+    const { hi, hv } = coreClaim.hiHv();
+    const claimHash = poseidon.hash([hi, hv]);
+    const bjjValid = pubKey.verifyPoseidon(claimHash, proof.signature);
+
+    if (!bjjValid) {
+      throw new Error('signature proof not valid');
+    }
+    await this.validateDIDDocumentAuth(
+      proof.issuerData.id,
+      resolverURL,
+      proof.issuerData.state.value
+    );
+
+    const credStatusType = proof.issuerData.credentialStatus.type;
+    const credStatusResolver = await credStatusResolverRegistry.get(credStatusType);
+    if (!credStatusResolver) {
+      throw new Error(`please register credential status resolver for ${credStatusType} type`);
+    }
+    const credStatus = await credStatusResolver.resolve(proof.issuerData.credentialStatus, {
+      issuerDID: proof.issuerData.id,
+      userDID: userDID
+    });
+    const stateValid = this.validateTreeState(credStatus.issuer);
+    if (!stateValid) {
+      throw new Error('invalid tree state');
+    }
+
+    const revocationNonce = BigInt(proof.issuerData.credentialStatus.revocationNonce || 0);
+    const proofValid = await verifyProof(
+      Hash.fromHex(credStatus.issuer.revocationTreeRoot),
+      credStatus.mtp,
+      revocationNonce,
+      BigInt(0)
+    );
+    if (!proofValid) {
+      throw new Error(`proof validation failed. revNonce=${revocationNonce}`);
+    }
+    if (credStatus.mtp.existence) {
+      throw new Error('credential is revoked');
+    }
+    return true;
+  }
+
+  private async verifyIden3SparseMerkleTreeProof(
+    proof: Iden3SparseMerkleTreeProof,
+    coreClaim: Claim,
+    resolverURL: string
+  ): Promise<boolean> {
+    await this.validateDIDDocumentAuth(
+      proof.issuerData.id,
+      resolverURL,
+      proof.issuerData.state.value
+    );
+    // root from proof == issuerData.state.сlaimsTreeRoot
+    const { hi, hv } = coreClaim.hiHv();
+    const rootFromProofValue = await rootFromProof(proof.mtp, hi, hv);
+    if (!rootFromProofValue.equals(proof.issuerData.state.claimsTreeRoot)) {
+      throw new Error('root from proof not equal to issuer data claims tree root');
```

Comment:
verifyIden3SparseMerkleTreeProof: root from proof not equal to issuer data claims tree root